### ISSUE TITLE
**UP** ignores certain function supertypes

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#1852]
Consider the following program:

```dart
void f1(int i, [bool b = false]) => print('f1: $i');
void f2(int i, {String s = ''}) => print('f2: $i');

var b = true;

void main() {
  (b? f1 : f2)(4.66920161);
}
```

The invocation of the conditional expression in main is allowed according to [the specification](https://github.com/dart-lang/language/blob/master/resources/type-system/upper-lower-bounds.md#upper-bounds), because the **UP** of the two function types is `Function`, in spite of the fact that `void Function(int)` is a common supertype of the two operands.

The CFE (via `dart` from commit 497a331365164dd70decfd8cc166d99b284e8d23) follows the specification, accepts the program without reporting any compile-time errors, and throws at run time.

However, the analyzer rejects the program and reports an error mentioning that it has computed `void Function(int)` as the **UP** of the two types.

We could add an extra case to the rules defining **UP** in order to obtain the improved upper bound in this case:

- **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k)`,
         `T1 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1l, Named)`) =
   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2l)` if:
     - None of the parameters in `Named` are required
     - Each `B0i` and `B1i` are equal types (syntactically)
     - `l` <= `k`
     - `R0` is **UP**(`T0`, `T1`)
     - `B2i` is `B0i`
     - The type of `P2i` is **DOWN**(`typeof(P0i)`, `typeof(P1i)`), and `P2i` is a required positional parameter.

There would also be a symmetric rule. If we are willing to break the style for brevity, we could use this:

- **UP**(`T1 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1l, Named)`,
         `T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k)`) =
   **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k)`,
         `T1 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1l, Named)`).

The new rules could be placed anywhere in the group of rules about **UP** on function types with explicitly specified parameter lists, because they are all mutually exclusive.

We could also introduce a small generalization to handle the following kind of situation:

```dart
void f1(int i, int j, [int k = 0, bool b = false]) => print('f1: $i');
void f2(int i, [int j = 0, int k = 0, String s = '']) => print('f2: $i');

var b = true;

void main() {
  (b? f1 : f2)(1, 2, 3, throw 0);
  (b? f1 : f2)(4, 5, 6);
}
```

In this case, the two function types have a common supertype which is `void Function(int, int, [int, Never])`, but both the CFE and the analyzer behave in a way which shows that they compute `Function` as the upper bound of `f1` and `f2`.

We could change the rule about function types with only positional parameters as follows:

- **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k)`,
         `T1 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1l)`) =
   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2q)` if:
     - each `B0i` and `B1i` are equal types (syntactically)
     - `q` is min(`k`, `l`)
     - `R0` is **UP**(`T0`, `T1`)
     - `B2i` is `B0i`
     - The type of `P2i` is **DOWN**(`typeof(P0i)`, `typeof(P1i)`), and `P2i` is optional iff both `P0i` and `P1i` is optional.

We may wish to cut off remaining parameters starting with the first one whose type is `Never`, but the current approach is to keep them, so we may also choose to keep them, just because it is in a narrow sense a breaking change to omit them.

I'd recommend that we adopt the first generalization, because the analyzer already supports it (and removing it from the analyzer would remove a compile-time error from certain call sites which are actually statically known to be type incorrect). For the second generalization, we don't have current behaviors to justify adopting it, but it would still have the same property: It would turn expressions of type `Function` into expressions that have a more informative function type, hence catching some type errors at compile time rather than at run time.

@natebosch, @leafpetersen, @munificent, @lrhn, @jakemac53, @stereotype441, WDYT?

**Comment by stereotype441** [comment]
Yeah, I would definitely support making the first change (using **DOWN** to unify parameter types), and I think that since the analyzer already does this, there's no need to make the change conditioned on a language version.

I would love to see us make the second change as well (allowing required positional and optional positional parameter types to be unified), but since it would be a change to both the analyzer and the CFE, I'd want us to do more investigation to figure out how breaking it would be in practice, so that we can decide whether or not to condition it on a language version.

@eernstg have you investigated whether there are similar issues with **DOWN**?

**Comment by leafpetersen** [comment]
Unifying the CFE and analyzer behavior is always a good thing.  Beyond that, I'm skeptical of the user value in investing in improvements to the UP and DOWN algorithm here, particularly relative to other known infelicities.

**Comment by eernstg** [comment]
> whether there are similar issues with **DOWN**?

There couldn't be a **DOWN** rule where a non-`Never` subtype is obtained from the mixed case (one function type has >0 optional positional parameters, the other one has >0 named parameters), so we can't add any rules that are similar to 'generalization 1'.

On the other hand, I think it's fair to say that **DOWN** already uses the approach proposed as generalization 2: For function types without named parameters, it does not require that the two function types have the same number of required positional parameters, and it computes the status (required/optional) for the resulting parameter `P2i` based on the requiredness of the corresponding `P0i` and `P1i`.
